src/core/reactor.cc

```diff
@@ -3544,6 +3544,12 @@ smp_options::smp_options(program_options::option_group* parent_group)
 {
 }
 
+thread_local metrics::impl::metric_implementations metric_impls;
+
+metrics::impl::metric_implementations& metrics::impl::get_metric_implementations() {
+    return metric_impls;
+}
+
```

Comment:
You mean you don't understand why it's declared here, or you don't understand why it's not `static`?

It is declared here because `metric_impls` needs to be alive during the reactor (declared [here](https://github.com/redpanda-data/seastar/blob/0fd17f4cb57365c4a340fbb4435fcc903871512a/src/core/reactor.cc#L3597)) cleanup as metrics are [de-registered](https://github.com/redpanda-data/seastar/blob/master/src/core/metrics.cc#L255) at that point. `thread_local` destructors are called in reverse order of initialisation, so `metric_impls` needs to initialise ahead of the [reactor_holder](https://github.com/redpanda-data/seastar/blob/0fd17f4cb57365c4a340fbb4435fcc903871512a/src/core/reactor.cc#L3597).

It's not static as `metric_impls` needs external linkage.

Reply:
I mean I don't know why its previous location as a static local had a problem.

But if it's an uninitialization order fiasco, then ok.